relay-metrics/src/aggregation.rs

```diff
@@ -1164,6 +1289,75 @@ impl fmt::Debug for CostTracker {
     }
 }
 
+/// An iterator returning batches of buckets fitting into a size budget.
+///
+/// The size budget is given through `max_flush_bytes`, though this is an approximate number. On
+/// every iteration, this iterator returns a `Vec<Bucket>` which serializes into a buffer of the
+/// specified size. Buckets at the end of each batch may be split to fit into the batch.
+///
+/// Since this uses an approximate function to estimate the size of buckets, the actual serialized
+/// payload may exceed the size. The estimation function is built in a way to guarantee the same
+/// order of magnitude.
+struct CappedBucketIter {
+    buckets: std::vec::IntoIter<Bucket>,
+    next_bucket: Option<Bucket>,
+    max_flush_bytes: usize,
+}
+
+impl CappedBucketIter {
+    pub fn new(buckets: Vec<Bucket>, max_flush_bytes: usize) -> Self {
+        let mut buckets = buckets.into_iter();
+        let next_bucket = buckets.next();
+
+        Self {
+            buckets,
+            next_bucket,
+            max_flush_bytes,
+        }
+    }
+}
+
+impl Iterator for CappedBucketIter {
+    type Item = Vec<Bucket>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut current_batch = Vec::new();
+        let mut remaining_bytes = self.max_flush_bytes;
+
+        while let Some(bucket) = self.next_bucket.take() {
+            let bucket_size = bucket.estimated_size();
+            if bucket_size <= remaining_bytes {
+                // the bucket fits
+                remaining_bytes -= bucket_size;
+                current_batch.push(bucket);
+                self.next_bucket = self.buckets.next();
+            } else if bucket_size < self.max_flush_bytes / BUCKET_SPLIT_FACTOR {
+                // the bucket is too small to split, move it entirely
+                self.next_bucket = Some(bucket);
+                break;
+            } else {
+                // the bucket is big enough to split
+                let (left, right) = bucket.split_at(remaining_bytes);
+                if let Some(left) = left {
+                    current_batch.push(left);
+                }
+
+                self.next_bucket = right;
+                break;
+            }
+        }
+
+        if current_batch.is_empty() {
+            self.next_bucket = None; // fuse
+            None
```

Comment:
This could silently lose data isn't it?  Should this not at least check if there still is a next bucket to process and log a sentry error?

Reply:
https://github.com/getsentry/relay/pull/1331